web/pandas/pdeps/0014-string-dtype.md

```diff
@@ -0,0 +1,373 @@
+use storage-specific options. Users are expected to specify a generic name (such
+as `"str"` or `"string"`), and that will give them their default string dtype
+(which depends on whether PyArrow is installed or not).
+
+For the generic string alias to specify the dtype, `"string"` is already used
+for the `StringDtype` using `pd.NA`. This PDEP proposes to use `"str"` for the
+new default `StringDtype` using `NaN`. This ensures backwards compatibility for
+code using `dtype="string"`, and was also chosen because `dtype="str"` or
+`dtype=str` currently already works to ensure your data is converted to
+strings (only using object dtype for the result).
+
+But for testing purposes and advanced use cases that want control over the exact
+variant of the `StringDtype`, we need some way to specify this and distinguish
+them from the other string dtypes.
+
+Currently (pandas 2.2), `StringDtype(storage="pyarrow_numpy")` is used for the new variant using `NaN`,
+where the `"pyarrow_numpy"` storage was used to disambiguate from the existing
+`"pyarrow"` option using `pd.NA`. However, `"pyarrow_numpy"` is a rather confusing
+option and doesn't generalize well. Therefore, this PDEP proposes a new naming
+scheme as outlined below, and `"pyarrow_numpy"` will be deprecated and removed
+before pandas 3.0.
+
+The `storage` keyword of `StringDtype` is kept to disambiguate the underlying
+storage of the string data (using pyarrow or python objects), but an additional
+`na_value` is introduced to disambiguate the the variants using NA semantics
+and NaN semantics.
+
+Overview of the different ways to specify a dtype and the resulting concrete
+dtype of the data:
+
+| User specification                          | Concrete dtype                                                | String alias                          | Note     |
+|---------------------------------------------|---------------------------------------------------------------|---------------------------------------|----------|
+| Unspecified (inference)                     | `StringDtype(storage="pyarrow"\|"python", na_value=np.nan)`   | "str"                                 | (1)      |
+| `"str"` or `StringDtype(na_value=np.nan)`   | `StringDtype(storage="pyarrow"\|"python", na_value=np.nan)`   | "str"                                 | (1)      |
+| `StringDtype("pyarrow", na_value=np.nan)`   | `StringDtype(storage="pyarrow", na_value=np.nan)`             | "str"                                 |          |
+| `StringDtype("python", na_value=np.nan)`    | `StringDtype(storage="python", na_value=np.nan)`              | "str"                                 |          |
+| `StringDtype("pyarrow")`                    | `StringDtype(storage="pyarrow", na_value=pd.NA)`              | "string[pyarrow]"                     |          |
+| `StringDtype("python")`                     | `StringDtype(storage="python", na_value=pd.NA)`               | "string[python]"                      |          |
```

Comment:
Can a user specify `"string[pyarrow]"` and `"string[python]"` as the `dtype` as it works in 2.2 ?

Reply:
Yes, nothing changes there (it's mentioned in one of the paragraphs below the table that in theory we could also stop supporting the storage in the string alias for those, but that's out of scope for the PDEP). The fact that they are listed in the "String alias" column is meant to indicate that you can use that string as a string alias when specifying a dtype.

Reply:
I think I am confused by the "string alias" aspect.  There are string aliases for specifying the dtype, but also string aliases for representing the dtype (i.e., what you see if you do `Series.dtype`.  I'm not sure that we are completely bidirectional in all of our dtypes with respect to the strings.  

Also, if we are saying that `"str"` will not indicate the storage type, should we not then deprecate the usage of `"string[python]"` and `"string[pyarrow]"` and just make it `"string"` ?  The asymmetry between `"str"` not specifying storage and `"string"` doesn't feel right.

Reply:
> I think I am confused by the "string alias" aspect. There are string aliases for specifying the dtype, but also string aliases for representing the dtype (i.e., what you see if you do `Series.dtype`. I'm not sure that we are completely bidirectional in all of our dtypes with respect to the strings.

I know, and I had been going back and forth in my draft calling this columm "String alias" vs "Dtype repr", because both are partly overlapping concepts, but then also "dtype repr" is confusing because there is both `__repr__` and `__str__` (eg for `__str__` (used in `df.dtypes`) we always just "string" without the storage suffix).

In the end, this column actually shows the dtype `__repr__`, but which in practice is the most specific string alias one can use to specify the dtype (i.e. for "string[pyarrow]" you can also use "string" if pyarrow is the default storage, but the table already includes too much content).



> Also, if we are saying that `"str"` will not indicate the storage type, should we not then deprecate the usage of `"string[python]"` and `"string[pyarrow]"` and just make it `"string"` ?

As mentioned in my previous comment, a paragraph is included to say this is left for a separate discussion (this would introduce a backwards incompatible change for existing users, while otherwise are not any on this front, and the exact way to indicate backends in string aliases, I would like to defer any final decision about that to the PDEP on logical types)

Reply:
> In the end, this column actually shows the dtype `__repr__`, but which in practice is the most specific string alias one can use to specify the dtype (i.e. for "string[pyarrow]" you can also use "string" if pyarrow is the default storage, but the table already includes too much content).

Maybe relabel "String alias" to "String alias for dtype input" ?

Part of the confusion is this with 2.2:
```python
>>> pd.Series(["abc"], dtype="string[pyarrow]")
0    abc
dtype: string
>>> pd.Series(["abc"], dtype="string[pyarrow]").dtype
string[pyarrow]
>>> pd.Series(["abc"], dtype="string")
0    abc
dtype: string
>>> pd.Series(["abc"], dtype="string").dtype
string[python]
```

When you print a `Series`, it shows the `dtype` as `"string"`, but if you look at the `dtype`, it shows `"string[python]"` or `"string[pyarrow]"`   This is where my confusion is coming from.

Reply:
> When you print a `Series`, it shows the `dtype` as `"string"`, but if you look at the `dtype`, it shows `"string[python]"` or `"string[pyarrow]"` This is where my confusion is coming from.

Yes, that's the difference between `__repr__` and `__str__` that I mentioned. Now, this is all existing behaviour, but so for the new dtype I am proposing to _not_ have any difference here (both repr and str would show `"str"`)

Reply:
I believe showing the storage in both `__repr__` and `__str__`  is very important. When you work of large Series, you have two main constraints: storage and computation efficiency. python backed and pyarrow backed are much different in these aspects. Whereas pyarrow is the best option most of the time, if you need mutability, python backed strings could be a better option.

Reply:
@arnaudlegout it's definitely true that there are trade-offs in deciding which option is best (and for now, you will always be able to specify the storage and inspect it through an attribute).

I think the intent at the moment, though, is to consider the default of pyarrow (if installed) as the general default that _almost_ everyone is expected to use, and that we don't incentivize too much to switch to the non-default storage (i.e. that it is more considered as an implementation detail which storage is used, for most users). 
At least that's the sense I got from the discussion here. But more real-world feedback once we actually switch the default in 3.0 will probably give better insights in how big the drawbacks are in workflows heavy on mutability, and to what extent we should more clearly document those trade-offs (at some point we will also have to decide whether to keep the object-dtype based fallback long term, for example when making pyarrow required or when having a numpy string dtype based one).

My personal take on showing the storage in the repr/str is that for _most_ users, I expect that they shouldn't have to care about the exact storage (or even be aware of the concept), and therefore I think it is more confusing to show this by default in the `df.dtypes` or `ser` repr (which both uses the dtype's `__str__`). 
If we do want to show it in the dtype's `__repr__` (to make it more discoverable), I would personally advocate for making the `__repr__` more like a class repr instead of a string alias (eg printing something like `<pandas.StringDtype(storage=...)>` instead of `string[storage]`, to avoid the `[]` use for this).

What exactly the `__repr__` should look like for this new dtype is not included exactly in the PDEP, so feel free to open a new issue to argue for including the storage where we can continue this discussion.

Reply:
done https://github.com/pandas-dev/pandas/issues/59342